### ISSUE TITLE
DDP should provide an option for not touching grad of locally unused params

### ISSUE THREAD

**Issue by baobablyh** [pytorch/pytorch#41101]
## 🐛 Bug

For multi task training with different ranks running different tasks, parameters of inactive tasks are unused locally, even though used globally. By default DDP would allreduce these parameters and update all with the mean of grad across all ranks. Such grad update could trigger the weight decay with some optimizers (such as SGD) even for those inactive tasks on rank. Such weight update would propagate to ranks running active tasks and eventually pollute the grad there. 

## To Reproduce

1. Have two ranks R1 and R2. Both of them load the same model.
1. R1 runs task A branch and R2 runs task B branch.
1. After the first iteration, we would see B's grad on R1 and A's grad on R2 get updated.

## Expected behavior

B's grad on R1 and A's grad on R2 should never get updated as task B is inactive on R1 and task A is inactive on R2.

**Comment by baobablyh** [comment]
@mrshenli

**Comment by mruberry** [comment]
Thank you for filing this issue, @baobablyh! I'll let @mrshenli comment on the proposal itself.

**Comment by mrshenli** [comment]
Hey @baobablyh, thanks for the request. 

The use case makes sense to me, but I am not sure if this should be added to DDP, as this will break DDP's invariant that it will keep the wrapped model synchronized across all processes. And thing can go wrong if, say after a few iterations, R1 starts to run task B and R2 starts to run task A. 

If the input model has a common components across all tasks, will it work if we just wrap that common part with DDP, and leave task-specific components as local models? If you also need to sync the task-specific part across a subset of the ranks, can they be wrapped with a different DDP instance operating on that subgroup?


Just curious, after training, how do you collect and merge the models from R1 and R2 if they are diverged?

**Comment by MagicFrogSJTU** [comment]
@baobablyh 
Is this your case?
There are multi-class heads, and in a DDP epoch, only 3 of them are activated. But during loss.backwards(), weight decay of all heads are triggered. However, in non-DDP mode, only 3 heads should be in the loss of weight decay.

**Comment by MagicFrogSJTU** [comment]
@mrshenli 
I look through the cpp code.
If a variable's grad is globally None, its grad will remain None in every replica. Am I right?

**Comment by mrshenli** [comment]
@MagicFrogSJTU  

> If a variable's grad is globally None, its grad will remain None rather than zero in every replica. Am I right?

Yep, this is correct. @baobablyh authored that part of code in #28883. I guess that's still not enough for the multi-class heads use case discussed above, due to 1) weight delay 2) grads for all params, even if they are globally unused, are still communicated using `AllReduce` in the backward pass, so DDP still pays the comm overhead. I curious to know if this can be solved by just wrapping part of the model with DDP.

**Comment by MagicFrogSJTU** [comment]
> @MagicFrogSJTU
> 
> > If a variable's grad is globally None, its grad will remain None rather than zero in every replica. Am I right?
> 
> Yep, this is correct. @baobablyh authored that part of code in #28883. I guess that's still not enough for the multi-class heads use case discussed above, due to 1) weight delay 2) grads for all params, even if they are globally unused, are still communicated using `AllReduce` in the backward pass, so DDP still pays the comm overhead. I curious to know if this can be solved by just wrapping part of the model with DDP.

Could you please tell me #28883 apply to which pytorch version? I am currently using pytorch1.5.

**Comment by mrshenli** [comment]
> Could you please tell me #28883 apply to which pytorch version? I am currently using pytorch1.5.

It is included in v1.5.0+

**Comment by baobablyh** [comment]
> Hey @baobablyh, thanks for the request.
> 
> The use case makes sense to me, but I am not sure if this should be added to DDP, as this will break DDP's invariant that it will keep the wrapped model synchronized across all processes. And thing can go wrong if, say after a few iterations, R1 starts to run task B and R2 starts to run task A.
> 
> If the input model has a common components across all tasks, will it work if we just wrap that common part with DDP, and leave task-specific components as local models? If you also need to sync the task-specific part across a subset of the ranks, can they be wrapped with a different DDP instance operating on that subgroup?
> 
> Just curious, after training, how do you collect and merge the models from R1 and R2 if they are diverged?

@mrshenli Thanks a lot for the reply which is very insightful.

I am not sure either if this should be added to DDP. That's why I have suggested maybe adding an option to support this special case :-)

Yes, for our case there is a common part shared across tasks. It would work as you suggested with one DDP for the common part and separate DDPs for task-specific parts. The thing is we also have use case for running task A and B at the same time in round robin fashion on R1 and R2 so maintaining these use cases together seems painful if we use single DDP for one and multiple DDPs for the other.

As for the model divergence, we sync periodically on each checkpoint.

**Comment by baobablyh** [comment]
> @MagicFrogSJTU
> 
> > If a variable's grad is globally None, its grad will remain None rather than zero in every replica. Am I right?
> 
> Yep, this is correct. @baobablyh authored that part of code in #28883. I guess that's still not enough for the multi-class heads use case discussed above, due to 1) weight delay 2) grads for all params, even if they are globally unused, are still communicated using `AllReduce` in the backward pass, so DDP still pays the comm overhead. I curious to know if this can be solved by just wrapping part of the model with DDP.

Yes, we have been handling two cases:
1. Run task A and B in round robin fashion on R1 and R2. The same task is active on all ranks at any given time. #28883 addressed this case.
2. Run A on R1 and run B on R2. This is the case we are trying to address with this ticket.

**Comment by mrshenli** [comment]
> The thing is we also have use case for running task A and B at the same time in round robin fashion on R1 and R2 so maintaining these use cases together seems painful if we use single DDP for one and multiple DDPs for the other.

I see. One possibility might be having a `ddp.ignore_none_grad()` context manager to wrap the iterations that should ignore local non grads. And it can also force model sync across all processes on exit. As this is a separate feature, this won't have any impact on existing use cases. But the downside is that the comm still takes place, and it just DDP ignores the communicated results. 

cc @zhaojuanmao

**Comment by zhaojuanmao** [comment]
the proposal about adding a context manager sounds good to me. @baobablyh I have a few noob questions:
1.  why not simply setting unused parameters with required_grad = false?  is it because these parameters could become used parameter from unused parameters during training?
2. if local/global unused parameters could change during training, is there a case like this?  is the result what you want?
         iteration 1, parameter 1 on rank 0 is unused, grad is None and not updated;
         iteration 2, parameter 1 on rank 0 is used, grad is set and updated;
         iteration 3, parameter 1 on rank0 is unused, grad will not be updated, but will be kept as zero, because grad has been reset as zero before backward.

**Comment by mrshenli** [comment]
Hey @baobablyh if the context manager idea can support your use case, @zhaojuanmao and I can bring this up to the team in our next meeting. And if there is no objection from the team, we can add a pitch to describe how this can be implemented.

**Comment by MagicFrogSJTU** [comment]
> the proposal about adding a context manager sounds good to me. @baobablyh I have a few noob questions:
> 
> 1. why not simply setting unused parameters with required_grad = false?  is it because these parameters could become used parameter from unused parameters during training?
> 2. if local/global unused parameters could change during training, is there a case like this?  is the result what you want?
>    iteration 1, parameter 1 on rank 0 is unused, grad is None and not updated;
>    iteration 2, parameter 1 on rank 0 is used, grad is set and updated;
>    iteration 3, parameter 1 on rank0 is unused, grad will not be updated, but will be kept as zero, because grad has been reset as zero before backward.

In the source code
```c++
void Optimizer::zero_grad() {
  for (auto& group : param_groups_) {
    for (auto& p : group.params()) {
      if (p.grad().defined()) {
        p.grad().detach_();
        p.grad().zero_();
      }
    }
  }
}
```
In the single-gpu (not DDP) mode, if a grad is once set, it will always exists. In the later iterations, it will be zero rather than None when it is unused. And this applies to DDP too.
Am I right?
@mrshenli

**Comment by baobablyh** [comment]
> Hey @baobablyh if the context manager idea can support your use case, @zhaojuanmao and I can bring this up to the team in our next meeting. And if there is no objection from the team, we can add a pitch to describe how this can be implemented.

I feel such context manager might be able to provide more flexibility for this kind of use cases. However I don't feel that would be very convenient for my use case where we would like to ignore none grad for locally unused params for the entire training and sync model across ranks during ckpt. An init param for DistributedDataParallel would be even better to specify such ignorance for the entire life cycle of DDP. The context manager can be used for fine-grained control if needed.

The model sync can be tricky as we have to broadcast from one of those ranks running the active task. DDP may not be aware of such task ownership. I feel the straightforward way would be have DDP user handle model sync-up as it has all the information about which ranks run what tasks.

**Comment by baobablyh** [comment]
> the proposal about adding a context manager sounds good to me. @baobablyh I have a few noob questions:
> 
> 1. why not simply setting unused parameters with required_grad = false?  is it because these parameters could become used parameter from unused parameters during training?
> 2. if local/global unused parameters could change during training, is there a case like this?  is the result what you want?
>    iteration 1, parameter 1 on rank 0 is unused, grad is None and not updated;
>    iteration 2, parameter 1 on rank 0 is used, grad is set and updated;
>    iteration 3, parameter 1 on rank0 is unused, grad will not be updated, but will be kept as zero, because grad has been reset as zero before backward.

In my case used/unused params stay unchanged during the training. It would not help if setting unused parameters with required_grad = false, as the grad of locally unused param get update after backward. Please see my temporary fix here: https://github.com/pytorch/pytorch/pull/41104/files#diff-c269190a925a4b0df49eda8a8f6c5bd3L674

**Comment by mrshenli** [comment]
@MagicFrogSJTU 

> In the single-gpu (not DDP) mode, if a grad is once set, it will always exists. In the later iterations, it will be zero rather than None when it is unused. And this applies to DDP too.

Yes, this is true. Apps have to manually reset the grads to None if necessary.

**Comment by mrshenli** [comment]
> I feel such context manager might be able to provide more flexibility for this kind of use cases. However I don't feel that would be very convenient for my use case where we would like to ignore none grad for locally unused params for the entire training and sync model across ranks during ckpt. An init param for DistributedDataParallel would be even better to specify such ignorance for the entire life cycle of DDP. The context manager can be used for fine-grained control if needed.

> The model sync can be tricky as we have to broadcast from one of those ranks running the active task. DDP may not be aware of such task ownership. I feel the straightforward way would be have DDP user handle model sync-up as it has all the information about which ranks run what tasks.

A few concerns here:

1. We are kind of reluctant to introduce more and more ctor arguments to DDP, as there are already 9 args, and we know there are quite a few projects that would like to add dedicated args for their use cases as well. We are trying to say no to them, because a) some of them cannot co-exists b) people might be confused about how to set those args when they call the ctor c) this will be a never-ending process. It will be great if this can be done outside of the ctor. One possibility is to provide a method (instead of a ctx manager), e.g., after calling `ddp.ignore_none_grad()`, the DDP inistance will set a parameter, and from there on it will ignore local args. 
2. This breaks DDP's assumption that gradients are always synchronized. I haven't thought through what are the implications here. E.g., what if I got a DDP instance from a 3rd-party lib, does it mean I have to understand this behavior and check if the `ignore_none_grad()` mode is enabled? If we cannot get this addressed, let's keep the param and method as "private", by naming it using a "_" prefix.

**Comment by zhaojuanmao** [comment]
> > the proposal about adding a context manager sounds good to me. @baobablyh I have a few noob questions:
> > 
> > 1. why not simply setting unused parameters with required_grad = false?  is it because these parameters could become used parameter from unused parameters during training?
> > 2. if local/global unused parameters could change during training, is there a case like this?  is the result what you want?
> >    iteration 1, parameter 1 on rank 0 is unused, grad is None and not updated;
> >    iteration 2, parameter 1 on rank 0 is used, grad is set and updated;
> >    iteration 3, parameter 1 on rank0 is unused, grad will not be updated, but will be kept as zero, because grad has been reset as zero before backward.
> 
> In my case used/unused params stay unchanged during the training. It would not help if setting unused parameters with required_grad = false, as the grad of locally unused param get update after backward. Please see my temporary fix here: https://github.com/pytorch/pytorch/pull/41104/files#diff-c269190a925a4b0df49eda8a8f6c5bd3L674

@baobablyh DDP will not sync and update params with require_grad = false. So if your app knows globally unused params and they will not change, you can freeze these params, DDP will not communicate them. 

For params that are unused on some ranks but could be used on other ranks, they have to be set as require_grad = true, otherwise DDP will have de-sync issue. So your issue still exists.

**Comment by baobablyh** [comment]
> > I feel such context manager might be able to provide more flexibility for this kind of use cases. However I don't feel that would be very convenient for my use case where we would like to ignore none grad for locally unused params for the entire training and sync model across ranks during ckpt. An init param for DistributedDataParallel would be even better to specify such ignorance for the entire life cycle of DDP. The context manager can be used for fine-grained control if needed.
> 
> > The model sync can be tricky as we have to broadcast from one of those ranks running the active task. DDP may not be aware of such task ownership. I feel the straightforward way would be have DDP user handle model sync-up as it has all the information about which ranks run what tasks.
> 
> A few concerns here:
> 
> 1. We are kind of reluctant to introduce more and more ctor arguments to DDP, as there are already 9 args, and we know there are quite a few projects that would like to add dedicated args for their use cases as well. We are trying to say no to them, because a) some of them cannot co-exists b) people might be confused about how to set those args when they call the ctor c) this will be a never-ending process. It will be great if this can be done outside of the ctor. One possibility is to provide a method (instead of a ctx manager), e.g., after calling `ddp.ignore_none_grad()`, the DDP inistance will set a parameter, and from there on it will ignore local args.

Make sense. I think it is very valid concern. Adding a method would work for me as well.

> 2. This breaks DDP's assumption that gradients are always synchronized. I haven't thought through what are the implications here. E.g., what if I got a DDP instance from a 3rd-party lib, does it mean I have to understand this behavior and check if the `ignore_none_grad()` mode is enabled? If we cannot get this addressed, let's keep the param and method as "private", by naming it using a "_" prefix.

Yeah, private method seems reasonable. We only need to use it for this special case. The user should fully understand how it works and take risk by themselves.

**Comment by baobablyh** [comment]
> > > the proposal about adding a context manager sounds good to me. @baobablyh I have a few noob questions:
> > > 
> > > 1. why not simply setting unused parameters with required_grad = false?  is it because these parameters could become used parameter from unused parameters during training?
> > > 2. if local/global unused parameters could change during training, is there a case like this?  is the result what you want?
> > >    iteration 1, parameter 1 on rank 0 is unused, grad is None and not updated;
> > >    iteration 2, parameter 1 on rank 0 is used, grad is set and updated;
> > >    iteration 3, parameter 1 on rank0 is unused, grad will not be updated, but will be kept as zero, because grad has been reset as zero before backward.
> > 
> > 
> > In my case used/unused params stay unchanged during the training. It would not help if setting unused parameters with required_grad = false, as the grad of locally unused param get update after backward. Please see my temporary fix here: https://github.com/pytorch/pytorch/pull/41104/files#diff-c269190a925a4b0df49eda8a8f6c5bd3L674
> 
> @baobablyh DDP will not sync and update params with require_grad = false. So if your app knows globally unused params and they will not change, you can freeze these params, DDP will not communicate them.
> 
> For params that are unused on some ranks but could be used on other ranks, they have to be set as require_grad = true, otherwise DDP will have de-sync issue. So your issue still exists.

I feel we may not be on the same page. The use case does need such de-sync as we discussed above. Regardless of DDP communicating those param or not, the issue is that current DDP sets none grad to zero grad anyway and the optimizer decays zero grade afterwards even if those params are not used at all on that rank. Here is the code doing the weight decay: https://github.com/pytorch/pytorch/blob/master/torch/optim/sgd.py#L95  The point is to keep none grad untouched so it does not trigger weight update.

**Comment by zhaojuanmao** [comment]
@baobablyh sorry I did not reply it clearly. 

1. I understand your issue, and try to echo that setting require_grad = false does not help your issue (not update local unused parameters) here. it does not only resolve your issue and also could cause de-sync/hang

2. but you mentioned your global unused params will not change during whole training, I'm wondering whether you can set these unused params with require_grad = false.  in this case, DDP will not communicate these global unused params and improve performance.  ==> This is separate discussion though.

**Comment by baobablyh** [comment]
> @baobablyh sorry I did not reply it clearly.
> 
> 1. I understand your issue, and try to echo that setting require_grad = false does not help your issue (not update local unused parameters) here. it does not only resolve your issue and also could cause de-sync/hang
> 2. but you mentioned your global unused params will not change during whole training, I'm wondering whether you can set these unused params with require_grad = false.  in this case, DDP will not communicate these global unused params and improve performance.  ==> This is separate discussion though.

Ah so sorry for misreading your reply. I got your point now. Yes not communicating global unused params would be a nice optimization for performance.

BTW, do you mind pointing me to the code place which excludes params with require_grad = false from DDP reduction? I thought we always have a spot in bucket even for unused param and communicate the flattened tensor as a whole.

**Comment by zhaojuanmao** [comment]
during ddp constructor, if parameter.require_grad = false, it will be filtered out and will not be passed to DDP reducer

**Comment by mrshenli** [comment]
> BTW, do you mind pointing me to the code place which excludes params with require_grad = false from DDP reduction? I thought we always have a spot in bucket even for unused param and communicate the flattened tensor as a whole.

I am not sure if this can solve the issue. Because if you do the above, all DDP instances need to set `require_grad = false` on the same set of parameters, otherwise, different DDP instances might end up with different buckets, causing AllReduce to fail. IIUC, in your use case, R1 and R2 will need to set `require_grad = false` on a different set of parameters?

**Comment by baobablyh** [comment]
> > BTW, do you mind pointing me to the code place which excludes params with require_grad = false from DDP reduction? I thought we always have a spot in bucket even for unused param and communicate the flattened tensor as a whole.
> 
> I am not sure if this can solve the issue. Because if you do the above, all DDP instances need to set `require_grad = false` on the same set of parameters, otherwise, different DDP instances might end up with different buckets, causing AllReduce to fail. IIUC, in your use case, R1 and R2 will need to set `require_grad = false` on a different set of parameters?

No it does not. As @zhaojuanmao mentioned, that is only for optimizing DDP communication on globally unused parameters. Just like you pointed out, it wouldn't help locally unused parameters.
